lib/teleterm/gateway/gateway.go

```diff
@@ -116,29 +145,114 @@ func (g *Gateway) Close() error {
 
 // Serve starts the underlying ALPN proxy. Blocks until closeContext is canceled.
 func (g *Gateway) Serve() error {
-	g.Log.Info("Gateway is open.")
+	g.cfg.Log.Info("Gateway is open.")
 
 	if err := g.localProxy.Start(g.closeContext); err != nil {
 		return trace.Wrap(err)
 	}
 
-	g.Log.Info("Gateway has closed.")
+	g.cfg.Log.Info("Gateway has closed.")
+
+	return nil
+}
+
+func (g *Gateway) URI() uri.ResourceURI {
+	return g.cfg.URI
+}
+
+func (g *Gateway) SetURI(newURI uri.ResourceURI) {
+	g.cfg.URI = newURI
+}
+
+func (g *Gateway) TargetURI() string {
+	return g.cfg.TargetURI
+}
+
+func (g *Gateway) TargetName() string {
+	return g.cfg.TargetName
+}
+
+func (g *Gateway) Protocol() string {
+	return g.cfg.Protocol
+}
+
+func (g *Gateway) TargetUser() string {
+	return g.cfg.TargetUser
+}
+
+func (g *Gateway) TargetSubresourceName() string {
+	return g.cfg.TargetSubresourceName
+}
+
+func (g *Gateway) SetTargetSubresourceName(value string) {
+	g.cfg.TargetSubresourceName = value
+}
+
+func (g *Gateway) Log() *logrus.Entry {
+	return g.cfg.Log
+}
+
+func (g *Gateway) LocalAddress() string {
+	return g.cfg.LocalAddress
+}
+
+func (g *Gateway) LocalPort() string {
+	return g.cfg.LocalPort
+}
+
+// SetLocalPortAndRestart attempts to create a listener on the specified port. If successful, it
+// obtains a listener on the new port, stops the old proxy, updates the fields on gateway and then
+// starts the new proxy with the new listener. It starts the local proxy even if the previous one
+// wasn't started – SetLocalPortAndRestart is expected to be called after Serve.
+//
+// If it fails it's imperative that the current proxy is kept intact and the fields on gateway are
+// not changed. This way if the user attempts to change the port to one that cannot be obtained,
+// they're able to correct that mistake and choose a different port.
+//
+// SetLocalPortAndRestart is a noop if port is equal to the existing port.
+func (g *Gateway) SetLocalPortAndRestart(port string) error {
+	if port == g.cfg.LocalPort {
+		return nil
+	}
+
+	result, err := newListenerAndLocalProxy(*g.cfg, port)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	previousPort := g.cfg.LocalPort
+
+	g.closeCancel()
+	if err = g.localProxy.Close(); err != nil {
+		g.cfg.Log.WithError(err).Warnf("Failed to close the previous proxy on port %s.", previousPort)
+	}
+
+	g.cfg.LocalPort = result.LocalPort
+	g.localProxy = result.localProxy
+	g.closeCancel = result.closeCancel
+	g.closeContext = result.closeContext
+
+	go func() {
+		if err := g.Serve(); err != nil {
```

Comment:
it makes me uneasy that we just fire a `Serve()` and never wait for it to finish or consume the error, probably outside the scope of this PR but I think it's going to be difficult to gracefully handle errors and restarts. In teleport we have `RegisterCriticalFunc` so that if things like this fail, we can restart. This is why I also think it's not great that we let the original `Serve()` goroutine die and just transparently call a new one here, in case we ever want to have some error handling for these in teleterm.

Reply:
> This is why I also think it's not great that we let the original `Serve()` goroutine die and just transparently call a new one here, in case we ever want to have some error handling for these in teleterm.

How should this look like if I wanted to attempt to improve it in the future? Should there be a dedicated goroutine that exclusively manages the lifecycle of the gateway and restarts it on failure?